Mudlet/Mudlet
# Fix: Stop sound/music precisely for MCMP

#### Brief overview of PR changes/additions
To correct a symptom where more media was stopping than was intended using the `key` parameter, an issue was fixed with the code that handles Mud Sound Protocol (MSP), Mud Client Media Protocol (MCMP), and the Sound/Music Lua API. Reproducing the issue identified that the TMediaData element of reused TMediaPlayer(s) were not getting updated as intended. Better handling of the QList(s) that cache the TMediaPlayer(s) corrected the problem. A side benefit of implementing this fix will be better reuse of the TMediaPlayer(s), generating more efficiency in Mudlet.

#### Motivation for adding to Mudlet
Mudlet Discord user Blizzard#7841 reported the issue on the #help channel.

<img width="1055" alt="Screen Shot 2022-10-23 at 5 06 20 PM" src="https://user-images.githubusercontent.com/3058178/197418309-790c5661-6e3f-4709-8119-5a6f133abe34.png">

#### Other info (issues closed, discussion etc)
Release Notes: Corrected a symptom where more media was stopping than was intended using the `key` parameter with Mud Client Media Protocol (MCMP) and the Sound/Music API for Mudlet.

**Testing**

This is an edge case, so the scenario here is a bit quirky to reproduce the issue.

In this scenario, we will play space music with a `key` of `space-music` and weirder things music with a `key` of `weirder-things`. The `weirder-things` music will stop the `space-music`, which plays at a lower priority. We will restart the `weirder-things` music and then try to stop only the `space-music` (which _is not_ playing at the time). In the older version of Mudlet, we'll hear that the `weirder-things` music stops too when pulling level P the second time, which is not desired. A bug in reusing media players caused more music (all music in this case) to stop, when it is only intended to stop the space music.

The issue appears in 4.16.0, where all music stops on step 8 below:

1. Open Mudlet, login to StickMUD at stickmud.com 7670 secure
2. Go e;2u
3. pull lever N with gmcp
4. pull lever O with gmcp
5. pull lever P with gmcp
6. pull lever Q with gmcp
7. pull lever O with gmcp
8. pull lever P with gmcp
9. close Mudlet

Pulling level P should have only stopped `space-music`, but it stops the `weirder-things` music also.

The issue shows fixed with the PR version where the weirder things music keeps playing on step 7 below: 

1. Open Mudlet, relogin to StickMUD at stickmud.com 7760 secure
2. pull lever N with gmcp
3. pull lever O with gmcp
4. pull lever P with gmcp
5. pull lever Q with gmcp
6. pull lever O with gmcp
7. pull lever P with gmcp
8. pull lever Q with gmcp
9. close Mudlet

Success is when pulling lever P the second time does nothing.  This is the intended functionality.  Use lever Q to stop all the music.
